Allow const parameters in array sizes to be unified

varkor opened the pull request:
Fixes https://github.com/rust-lang/rust/issues/60632.
Fixes https://github.com/rust-lang/rust/issues/60744.
Fixes https://github.com/rust-lang/rust/pull/60923.
(The last commit should probably be viewed in isolation, as it just renames things from `type` to `kind`.)

r? @eddyb

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/199433977) ([raw log](https://api.travis-ci.com/v3/job/199433977/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:11ec65fe:start=1557598173129579784,finish=1557598258577645449,duration=85448065665
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---
travis_time:start:test_ui
Check compiletest suite=ui mode=ui (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:09:29] 
[01:09:29] running 5527 tests
[01:09:32] ....................................F............................................................... 100/5527
[01:09:41] .................................................................................................... 300/5527
[01:09:44] .................................................................................................... 400/5527
[01:09:44] .................................................................................................... 400/5527
[01:09:47] ...........................................................................F......................i. 500/5527
[01:09:55] .................................................................................................... 700/5527
[01:10:00] .................................................................................................... 800/5527
[01:10:05] ...............................................................................i...............i.... 900/5527
[01:10:09] .................................................................................................... 1000/5527
---
[01:12:36] .................................................................................................... 4900/5527
[01:12:39] .................................................................................................... 5000/5527
[01:12:44] .................................................................................................... 5100/5527
[01:12:48] .................................................................................................... 5200/5527
[01:12:51] .............F...................................................................................... 5300/5527
[01:12:57] .................................................................i.................................. 5500/5527
[01:12:58] ...........................
[01:12:58] failures:
[01:12:58] 
[01:12:58] 
[01:12:58] ---- [ui] ui/array-break-length.rs stdout ----
[01:12:58] diff of stderr:
[01:12:58] 
[01:12:58] 10 LL |         |_: [_; continue]| {}
[01:12:58] 11    |                 ^^^^^^^^ cannot break outside of a loop
[01:12:58] - error: aborting due to 2 previous errors
[01:12:58] + error[E0308]: mismatched types
[01:12:58] +   --> $DIR/array-break-length.rs:3:9
[01:12:58] +    |
[01:12:58] +    |
[01:12:58] + LL |         |_: [_; break]| {}
[01:12:58] +    |         ^^^^^^^^^^^^^^^^^^ expected (), found closure
[01:12:58] +    = note: expected type `()`
[01:12:58] +    = note: expected type `()`
[01:12:58] +               found type `[closure@$DIR/array-break-length.rs:3:9: 3:27]`
[01:12:58] - For more information about this error, try `rustc --explain E0268`.
[01:12:58] + error[E0308]: mismatched types
[01:12:58] +   --> $DIR/array-break-length.rs:7:9
[01:12:58] +    |
[01:12:58] +    |
[01:12:58] + LL |         |_: [_; continue]| {}
[01:12:58] +    |         ^^^^^^^^^^^^^^^^^^^^^ expected (), found closure
[01:12:58] +    = note: expected type `()`
[01:12:58] +    = note: expected type `()`
[01:12:58] +               found type `[closure@$DIR/array-break-length.rs:7:9: 7:30]`
[01:12:58] + error: aborting due to 4 previous errors
[01:12:58] + 
[01:12:58] + Some errors have detailed explanations: E0268, E0308.
[01:12:58] + For more information about an error, try `rustc --explain E0268`.
[01:12:58] + For more information about an error, try `rustc --explain E0268`.
[01:12:58] 16 
[01:12:58] 
[01:12:58] 
[01:12:58] The actual stderr differed from the expected stderr.
[01:12:58] Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/array-break-length/array-break-length.stderr
[01:12:58] To update references, rerun the tests and pass the `--bless` flag
[01:12:58] To only update this specific test, also pass `--test-args array-break-length.rs`
[01:12:58] error: 1 errors occurred comparing output.
[01:12:58] status: exit code: 1
[01:12:58] status: exit code: 1
[01:12:58] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/array-break-length.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/array-break-length" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/array-break-length/auxiliary" "-A" "unused"
[01:12:58] ------------------------------------------
[01:12:58] 
[01:12:58] ------------------------------------------
[01:12:58] stderr:
[01:12:58] stderr:
[01:12:58] ------------------------------------------
[01:12:58] error[E0268]: `break` outside of loop
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |         |_: [_; break]| {} //~ ERROR: `break` outside of loop
[01:12:58]    |                 ^^^^^ cannot break outside of a loop
[01:12:58] 
[01:12:58] error[E0268]: `continue` outside of loop
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |         |_: [_; continue]| {} //~ ERROR: `continue` outside of loop
[01:12:58]    |                 ^^^^^^^^ cannot break outside of a loop
[01:12:58] error[E0308]: mismatched types
[01:12:58]   --> /checkout/src/test/ui/array-break-length.rs:3:9
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |         |_: [_; break]| {} //~ ERROR: `break` outside of loop
[01:12:58]    |         ^^^^^^^^^^^^^^^^^^ expected (), found closure
[01:12:58]    = note: expected type `()`
[01:12:58]    = note: expected type `()`
[01:12:58]               found type `[closure@/checkout/src/test/ui/array-break-length.rs:3:9: 3:27]`
[01:12:58] error[E0308]: mismatched types
[01:12:58]   --> /checkout/src/test/ui/array-break-length.rs:7:9
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |         |_: [_; continue]| {} //~ ERROR: `continue` outside of loop
[01:12:58]    |         ^^^^^^^^^^^^^^^^^^^^^ expected (), found closure
[01:12:58]    = note: expected type `()`
[01:12:58]    = note: expected type `()`
[01:12:58]               found type `[closure@/checkout/src/test/ui/array-break-length.rs:7:9: 7:30]`
[01:12:58] error: aborting due to 4 previous errors
[01:12:58] 
[01:12:58] Some errors have detailed explanations: E0268, E0308.
[01:12:58] For more information about an error, try `rustc --explain E0268`.
[01:12:58] For more information about an error, try `rustc --explain E0268`.
[01:12:58] 
[01:12:58] ------------------------------------------
[01:12:58] 
[01:12:58] 
[01:12:58] ---- [ui] ui/closures/closure-array-break-length.rs stdout ----
[01:12:58] diff of stderr:
[01:12:58] 
[01:12:58] 16 LL |     while |_: [_; break]| {} {}
[01:12:58] 17    |                   ^^^^^ cannot break outside of a loop
[01:12:58] - error: aborting due to 3 previous errors
[01:12:58] + error[E0308]: mismatched types
[01:12:58] +   --> $DIR/closure-array-break-length.rs:4:11
[01:12:58] +    |
[01:12:58] +    |
[01:12:58] + LL |     while |_: [_; continue]| {} {}
[01:12:58] +    |
[01:12:58] +    = note: expected type `bool`
[01:12:58] +    = note: expected type `bool`
[01:12:58] +               found type `[closure@$DIR/closure-array-break-length.rs:4:11: 4:32]`
[01:12:58] - For more information about this error, try `rustc --explain E0268`.
[01:12:58] + error[E0308]: mismatched types
[01:12:58] +   --> $DIR/closure-array-break-length.rs:6:11
[01:12:58] +    |
[01:12:58] +    |
[01:12:58] + LL |     while |_: [_; break]| {} {}
[01:12:58] +    |
[01:12:58] +    = note: expected type `bool`
[01:12:58] +    = note: expected type `bool`
[01:12:58] +               found type `[closure@$DIR/closure-array-break-length.rs:6:11: 6:29]`
[01:12:58] + error: aborting due to 5 previous errors
[01:12:58] + 
[01:12:58] + Some errors have detailed explanations: E0268, E0308.
[01:12:58] + For more information about an error, try `rustc --explain E0268`.
[01:12:58] + For more information about an error, try `rustc --explain E0268`.
[01:12:58] 22 
[01:12:58] 
[01:12:58] 
[01:12:58] The actual stderr differed from the expected stderr.
[01:12:58] Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/closures/closure-array-break-length/closure-array-break-length.stderr
[01:12:58] To update references, rerun the tests and pass the `--bless` flag
[01:12:58] To only update this specific test, also pass `--test-args closures/closure-array-break-length.rs`
[01:12:58] error: 1 errors occurred comparing output.
[01:12:58] status: exit code: 1
[01:12:58] status: exit code: 1
[01:12:58] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/closures/closure-array-break-length.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/closures/closure-array-break-length" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/closures/closure-array-break-length/auxiliary" "-A" "unused"
[01:12:58] ------------------------------------------
[01:12:58] 
[01:12:58] ------------------------------------------
[01:12:58] stderr:
[01:12:58] stderr:
[01:12:58] ------------------------------------------
[01:12:58] error[E0268]: `continue` outside of loop
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |     |_: [_; continue]| {}; //~ ERROR: `continue` outside of loop
[01:12:58]    |             ^^^^^^^^ cannot break outside of a loop
[01:12:58] 
[01:12:58] error[E0268]: `continue` outside of loop
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |     while |_: [_; continue]| {} {} //~ ERROR: `continue` outside of loop
[01:12:58]    |                   ^^^^^^^^ cannot break outside of a loop
[01:12:58] 
[01:12:58] error[E0268]: `break` outside of loop
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |     while |_: [_; break]| {} {} //~ ERROR: `break` outside of loop
[01:12:58]    |                   ^^^^^ cannot break outside of a loop
[01:12:58] error[E0308]: mismatched types
[01:12:58]   --> /checkout/src/test/ui/closures/closure-array-break-length.rs:4:11
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |     while |_: [_; continue]| {} {} //~ ERROR: `continue` outside of loop
[01:12:58]    |
[01:12:58]    = note: expected type `bool`
[01:12:58]    = note: expected type `bool`
[01:12:58]               found type `[closure@/checkout/src/test/ui/closures/closure-array-break-length.rs:4:11: 4:32]`
[01:12:58] error[E0308]: mismatched types
[01:12:58]   --> /checkout/src/test/ui/closures/closure-array-break-length.rs:6:11
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |     while |_: [_; break]| {} {} //~ ERROR: `break` outside of loop
[01:12:58]    |
[01:12:58]    = note: expected type `bool`
[01:12:58]    = note: expected type `bool`
[01:12:58]               found type `[closure@/checkout/src/test/ui/closures/closure-array-break-length.rs:6:11: 6:29]`
[01:12:58] error: aborting due to 5 previous errors
[01:12:58] 
[01:12:58] Some errors have detailed explanations: E0268, E0308.
[01:12:58] For more information about an error, try `rustc --explain E0268`.
[01:12:58] For more information about an error, try `rustc --explain E0268`.
[01:12:58] 
[01:12:58] ------------------------------------------
[01:12:58] 
[01:12:58] 
[01:12:58] ---- [ui] ui/type/type-dependent-def-issue-49241.rs stdout ----
[01:12:58] diff of stderr:
[01:12:58] 
[01:12:58] 10 LL |     let s: [u32; l] = v.into_iter().collect();
[01:12:58] 11    |                  ^ referenced constant has errors
[01:12:58] - error: aborting due to 2 previous errors
[01:12:58] - error: aborting due to 2 previous errors
[01:12:58] + error[E0277]: a collection of type `[u32; _]` cannot be built from an iterator over elements of type `{integer}`
[01:12:58] +    |
[01:12:58] +    |
[01:12:58] + LL |     let s: [u32; l] = v.into_iter().collect();
[01:12:58] +    |                                     ^^^^^^^ a collection of type `[u32; _]` cannot be built from `std::iter::Iterator<Item={integer}>`
[01:12:58] +    |
[01:12:58] +    = help: the trait `std::iter::FromIterator<{integer}>` is not implemented for `[u32; _]`
[01:12:58] - Some errors have detailed explanations: E0080, E0435.
[01:12:58] + error: aborting due to 3 previous errors
[01:12:58] + 
[01:12:58] + Some errors have detailed explanations: E0080, E0277, E0435.
[01:12:58] + Some errors have detailed explanations: E0080, E0277, E0435.
[01:12:58] 16 For more information about an error, try `rustc --explain E0080`.
[01:12:58] 17 
[01:12:58] 
[01:12:58] 
[01:12:58] The actual stderr differed from the expected stderr.
[01:12:58] Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/type-dependent-def-issue-49241/type-dependent-def-issue-49241.stderr
[01:12:58] To update references, rerun the tests and pass the `--bless` flag
[01:12:58] To only update this specific test, also pass `--test-args type/type-dependent-def-issue-49241.rs`
[01:12:58] error: 1 errors occurred comparing output.
[01:12:58] status: exit code: 1
[01:12:58] status: exit code: 1
[01:12:58] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/type/type-dependent-def-issue-49241.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/type-dependent-def-issue-49241" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/type/type-dependent-def-issue-49241/auxiliary" "-A" "unused"
[01:12:58] ------------------------------------------
[01:12:58] 
[01:12:58] ------------------------------------------
[01:12:58] stderr:
[01:12:58] stderr:
[01:12:58] ------------------------------------------
[01:12:58] error[E0435]: attempt to use a non-constant value in a constant
[01:12:58]   --> /checkout/src/test/ui/type/type-dependent-def-issue-49241.rs:3:22
[01:12:58]    |
[01:12:58] LL |     const l: usize = v.count(); //~ ERROR attempt to use a non-constant value in a constant
[01:12:58]    |                      ^ non-constant value
[01:12:58] error[E0080]: evaluation of constant value failed
[01:12:58]   --> /checkout/src/test/ui/type/type-dependent-def-issue-49241.rs:4:18
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |     let s: [u32; l] = v.into_iter().collect();
[01:12:58]    |                  ^ referenced constant has errors
[01:12:58] 
[01:12:58] error[E0277]: a collection of type `[u32; _]` cannot be built from an iterator over elements of type `{integer}`
[01:12:58]    |
[01:12:58]    |
[01:12:58] LL |     let s: [u32; l] = v.into_iter().collect();
[01:12:58]    |                                     ^^^^^^^ a collection of type `[u32; _]` cannot be built from `std::iter::Iterator<Item={integer}>`
[01:12:58]    |
[01:12:58]    = help: the trait `std::iter::FromIterator<{integer}>` is not implemented for `[u32; _]`
[01:12:58] error: aborting due to 3 previous errors
[01:12:58] 
[01:12:58] Some errors have detailed explanations: E0080, E0277, E0435.
[01:12:58] For more information about an error, try `rustc --explain E0080`.
---
[01:12:58] thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:512:22
[01:12:58] note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
[01:12:58] 
[01:12:58] 
[01:12:58] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:12:58] 
[01:12:58] 
[01:12:58] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:12:58] Build completed unsuccessfully in 0:04:44
[01:12:58] Build completed unsuccessfully in 0:04:44
[01:12:58] Makefile:48: recipe for target 'check' failed
[01:12:58] make: *** [check] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:068d1c50
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Sat May 11 19:24:05 UTC 2019
---
travis_time:end:0d2b9dc9:start=1557602646801958089,finish=1557602646809653344,duration=7695255
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:15e1e6b4
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:03b423d9
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

varkor commented:
This changes some test output, but it's in very rare cases, and the messages are not incorrect (they're just a little extraneous).

bors commented:
:umbrella: The latest upstream changes (presumably #60765) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented on src/librustc/ty/relate.rs:
```diff
@@ -493,36 +493,32 @@ pub fn super_relate_tys<'a, 'gcx, 'tcx, R>(relation: &mut R,
-                (Err(ErrorReported), _) | (_, Err(ErrorReported)) => {
-                    Ok(tcx.types.err)
+                _ => {
+                    if let Ok(sz) = relation.relate(&sz_a, &sz_b) {
```
Wait, why do we do anything other than this now?

eddyb commented on src/test/ui/consts/const-array-oob-arith.stderr:
```diff
@@ -11,7 +11,7 @@ error[E0308]: mismatched types
    |
 LL | const BOO: [i32; (ARR[0] - 41) as usize] = [5, 99];
-   |                                            ^^^^^^^ expected an array with a fixed size of 1 elements, found one with 2 elements
+   |                                            ^^^^^^^ expected `Const { ty: usize, val: Scalar(Bits { size: 8, bits: 1 }) }`, found `Const { ty: usize, val: Scalar(Bits { size: 8, bits: 2 }) }`
```
cc @oli-obk should we wait until printing this is fixed?

eddyb commented on src/test/ui/const-generics/uninferred-consts-during-codegen-1.rs:
```diff
@@ -0,0 +1,18 @@
+
+struct Array<T, const N: usize>([T; N]);
+
+impl<T: fmt::Debug, const N: usize> fmt::Debug for Array<T, {N}> {
```
`N` should also work, how are we tracking that?

varkor replied:
This is https://github.com/rust-lang/rust/issues/6080.

varkor commented:
Blocks https://github.com/rust-lang/rust/pull/60839.
@bors p=1

yodaldevoid commented:
I believe this is waiting on #59276, correct?

varkor commented:
@yodaldevoid: that is correct.

varkor commented:
I've added some tests to confirm this also fixes https://github.com/rust-lang/rust/pull/60742.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/201651749) ([raw log](https://api.travis-ci.com/v3/job/201651749/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:2c19a185:start=1558389884168661708,finish=1558389989483240960,duration=105314579252
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---
[01:09:47] .................................................................................................... 300/5536
[01:09:50] .................................................................................................... 400/5536
[01:09:53] .................................................................................................i.. 500/5536
[01:09:57] .................................................................................................... 600/5536
[01:10:01] .....................................................................F.............................. 700/5536
[01:10:11] ...................................................................................i...............i 900/5536
[01:10:15] .................................................................................................... 1000/5536
[01:10:19] ................iiiii............................................................................... 1100/5536
[01:10:21] .................................................................................................... 1200/5536
---
[01:13:03] 
[01:13:03] ---- [ui] ui/const-generics/cannot-infer-type-for-const-param.rs stdout ----
[01:13:03] diff of stderr:
[01:13:03] 
[01:13:03] 10 LL |     let _ = Foo::<3>([1, 2, 3]);
[01:13:03] 11    |                   ^ cannot infer type for `{integer}`
[01:13:03] - error: aborting due to previous error
[01:13:03] + error[E0308]: mismatched types
[01:13:03] +   --> $DIR/cannot-infer-type-for-const-param.rs:10:22
[01:13:03] +    |
[01:13:03] +    |
[01:13:03] + LL |     let _ = Foo::<3>([1, 2, 3]);
[01:13:03] +    |                      ^^^^^^^^^ expected `Const { ty: usize, val: Unevaluated(DefId(0:18 ~ cannot_infer_type_for_const_param[317d]::main[0]::{{constant}}[0]), []) }`, found `Const { ty: usize, val: Scalar(Bits { size: 8, bits: 3 }) }`
[01:13:03] +    |
[01:13:03] +    = note: expected type `[u8; _]`
[01:13:03] +               found type `[u8; 3]`
[01:13:03] - For more information about this error, try `rustc --explain E0282`.
[01:13:03] + error: aborting due to 2 previous errors
[01:13:03] + 
[01:13:03] + Some errors have detailed explanations: E0282, E0308.
[01:13:03] + Some errors have detailed explanations: E0282, E0308.
[01:13:03] + For more information about an error, try `rustc --explain E0282`.
[01:13:03] 16 
[01:13:03] 
[01:13:03] 
[01:13:03] The actual stderr differed from the expected stderr.
[01:13:03] Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/cannot-infer-type-for-const-param/cannot-infer-type-for-const-param.stderr
[01:13:03] To update references, rerun the tests and pass the `--bless` flag
[01:13:03] To only update this specific test, also pass `--test-args const-generics/cannot-infer-type-for-const-param.rs`
[01:13:03] error: 1 errors occurred comparing output.
[01:13:03] status: exit code: 1
[01:13:03] status: exit code: 1
[01:13:03] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/const-generics/cannot-infer-type-for-const-param.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/cannot-infer-type-for-const-param" "-Crpath" "-O" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/const-generics/cannot-infer-type-for-const-param/auxiliary" "-A" "unused"
[01:13:03] ------------------------------------------
[01:13:03] 
[01:13:03] ------------------------------------------
[01:13:03] stderr:
---
[01:13:03] 
[01:13:03] error[E0282]: type annotations needed
[01:13:03]   --> /checkout/src/test/ui/const-generics/cannot-infer-type-for-const-param.rs:10:19
[01:13:03]    |
[01:13:03] LL |     let _ = Foo::<3>([1, 2, 3]); //~ ERROR type annotations needed
[01:13:03]    |                   ^ cannot infer type for `{integer}`
[01:13:03] error[E0308]: mismatched types
[01:13:03]   --> /checkout/src/test/ui/const-generics/cannot-infer-type-for-const-param.rs:10:22
[01:13:03]    |
[01:13:03]    |
[01:13:03] LL |     let _ = Foo::<3>([1, 2, 3]); //~ ERROR type annotations needed
[01:13:03]    |                      ^^^^^^^^^ expected `Const { ty: usize, val: Unevaluated(DefId(0:18 ~ cannot_infer_type_for_const_param[317d]::main[0]::{{constant}}[0]), []) }`, found `Const { ty: usize, val: Scalar(Bits { size: 8, bits: 3 }) }`
[01:13:03]    |
[01:13:03]    = note: expected type `[u8; _]`
[01:13:03]               found type `[u8; 3]`
[01:13:03] error: aborting due to 2 previous errors
[01:13:03] 
[01:13:03] Some errors have detailed explanations: E0282, E0308.
[01:13:03] For more information about an error, try `rustc --explain E0282`.
---
[01:13:03] thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:512:22
[01:13:03] note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
[01:13:03] 
[01:13:03] 
[01:13:03] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:13:03] 
[01:13:03] 
[01:13:03] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:13:03] Build completed unsuccessfully in 0:04:43
[01:13:03] Build completed unsuccessfully in 0:04:43
[01:13:03] make: *** [check] Error 1
[01:13:03] Makefile:48: recipe for target 'check' failed
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:02163b5c
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Mon May 20 23:19:41 UTC 2019
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

varkor commented:
@eddyb: I've updated the tests following https://github.com/rust-lang/rust/pull/59276.

bjorn3 commented on src/test/ui/consts/const-array-oob-arith.stderr:
```diff
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
    |
 LL | const BLUB: [i32; (ARR[0] - 40) as usize] = [5];
-   |                                             ^^^ expected an array with a fixed size of 2 elements, found one with 1 elements
+   |                                             ^^^ expected `2usize`, found `1usize`
```
I prefer the old message. Could ot be special cased?

varkor commented:
I've added back in the special cased message for arrays of different lengths and fixed a typo in related diagnostics. This is ready for re-review @eddyb.

eddyb commented on src/test/ui/const-generics/cannot-infer-type-for-const-param.stderr:
```diff
@@ -14,7 +14,7 @@ error[E0308]: mismatched types
    |
 LL |     let _ = Foo::<3>([1, 2, 3]);
-   |                      ^^^^^^^^^ expected `Const { ty: usize, val: Unevaluated(DefId(0:18 ~ cannot_infer_type_for_const_param[317d]::main[0]::{{constant}}[0]), []) }`, found `Const { ty: usize, val: Scalar(0x0000000000000003) }`
+   |                      ^^^^^^^^^ expected `3`, found `3usize`
```
This... is really weird? I assume the `3` is @oli-obk's printing of same-crate anon consts?

varkor replied:
I think it's because it can't infer the specific type, rather than an issue with printing. This is a bug, though, which should be fixed by https://github.com/rust-lang/rust/pull/60839. (Which is ready to go as soon as this one is.)

eddyb replied:
I mean that the printing output is from `Unevaluated` for the `3` in `Foo::<3>` and not *the value* `3`.

oli-obk replied:
ugh, ok I did not envision user facing effects from that. So.... should I just do `_` or `{{unevaluated}}` or sth?

eddyb replied:
I don't think we need to jump to something else here, but we should consider it carefully.
`_` is clearly the conservative choice, as for "unevaluated" - I don't like that name too much, it just happened to be useful in `ty::Const` (and we could/should pick another name).

eddyb commented on src/librustc/infer/resolve.rs:
```diff
@@ -1,28 +1,28 @@
 
-/// The opportunistic type resolver can be used at any time. It simply replaces
-/// type variables that have been unified with the things they have
+/// The opportunistic kind resolver can be used at any time. It simply replaces
```
Here and above, still says "kind". Given that this is inside `infer` you can just omit the word.

eddyb commented on src/librustc/ty/error.rs:
```diff
@@ -94,17 +100,21 @@ impl<'tcx> fmt::Display for TypeError<'tcx> {
-            TupleSize(values) => {
-                write!(f, "expected a tuple with {} elements, \
-                           found one with {} elements",
+            FixedArraySize(values) => {
```
Why was the order reversed?

varkor replied:
It's this way round in the definition and everywhere else.

eddyb commented on src/librustc/ty/relate.rs:
```diff
@@ -474,55 +472,20 @@ pub fn super_relate_tys<'a, 'gcx, 'tcx, R>(relation: &mut R,
+                Err(err) => {
+                    // Check whether the lengths are both `usize`s,
+                    // but differ in value, for better diagnostics.
+                    // FIXME(eddyb) get the right param_env.
```
The `FIXME` is no longer relevant.

eddyb commented on src/librustc/ty/relate.rs:
```diff
@@ -474,55 +472,20 @@ pub fn super_relate_tys<'a, 'gcx, 'tcx, R>(relation: &mut R,
+            match relation.relate(&sz_a, &sz_b) {
+                Ok(sz) => Ok(tcx.mk_ty(ty::Array(t, sz))),
+                Err(err) => {
+                    // Check whether the lengths are both `usize`s,
```
Not just `usize` but *concrete*/*known* values.

varkor commented:
@bors r=eddyb

bors commented:
:pushpin: Commit c6bbb81b41ed0beca8b956fbf89cc9815c86ecb7 has been approved by `eddyb`

<!-- @bors r=eddyb c6bbb81b41ed0beca8b956fbf89cc9815c86ecb7 -->
<!-- homu: {"type":"Approved","sha":"c6bbb81b41ed0beca8b956fbf89cc9815c86ecb7","approver":"eddyb"} -->

Centril commented on src/test/ui/const-generics/uninferred-consts-during-codegen-2.rs:
```diff
@@ -0,0 +1,18 @@
+}
+
+fn main() {
+    println!("{:?}", Array([1, 2, 3]));
```
Why use `println!(...)` instead of `format!(...)`  + `assert_eq!(...)`?

Centril replied:
If actually testing the dynamic semantics isn't important then `println!(...)` shouldn't be used and you can just use an assertion function with a bound instead and have `// compile-pass`.

Centril commented on src/test/ui/const-generics/uninferred-consts-during-codegen-1.rs:
```diff
@@ -0,0 +1,18 @@
+}
+
+fn main() {
+    println!("{:?}", Array([1, 2, 3]));
```
Same here...

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout fn-const-array-parameter` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self fn-const-array-parameter --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
Auto-merging src/librustc_typeck/check/mod.rs
Auto-merging src/librustc_typeck/check/method/probe.rs
Auto-merging src/librustc_typeck/check/method/mod.rs
Auto-merging src/librustc_typeck/check/_match.rs
Auto-merging src/librustc/ty/relate.rs
CONFLICT (content): Merge conflict in src/librustc/ty/relate.rs
Auto-merging src/librustc/traits/specialize/mod.rs
Auto-merging src/librustc/traits/project.rs
Auto-merging src/librustc/traits/mod.rs
Auto-merging src/librustc/middle/mem_categorization.rs
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

varkor commented:
@bors r=eddyb

bors commented:
:pushpin: Commit 6233d1fee5df70258b02bb549967bc9319014081 has been approved by `eddyb`

<!-- @bors r=eddyb 6233d1fee5df70258b02bb549967bc9319014081 -->
<!-- homu: {"type":"Approved","sha":"6233d1fee5df70258b02bb549967bc9319014081","approver":"eddyb"} -->
